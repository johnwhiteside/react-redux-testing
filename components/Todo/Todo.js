import React from 'react';
import PropTypes from 'prop-types';

class Todo extends React.PureComponent {
	handleDeleteClick(){
		const { todo, onDeleteClick } = this.props;
		onDeleteClick(todo);
	}

	handleCheckboxClick(){
		const { todo, onCheckboxClick } = this.props;
		onCheckboxClick(todo);
	}

	render(){
		const { todo } = this.props;
		return (
			<div>
				<input checked={todo.isComplete} type="checkbox" onChange={this.handleCheckboxClick} />
				<span>{todo.label}</span>
				<span onClick={this.handleDeleteClick}> delete</span>
			</div>
		);
	}
};

Todo.propTypes = {
	todo: PropTypes.shape({
		isComplete: PropTypes.bool,
		label: PropTypes.string,
	}).isRequired,
	onCheckboxClick: PropTypes.func,
	onDeleteClick: PropTypes.func,
}

Todo.defaultProps = {
	onCheckboxClick: () => {},
	onDeleteClick: () => {},
}

export default Todo;
