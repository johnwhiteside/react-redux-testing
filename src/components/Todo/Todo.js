import React from 'react';
import PropTypes from 'prop-types';

class Todo extends React.PureComponent {
	constructor(props) {
		super(props);
		this.state = { value: '' }
		this.handleDeleteClick = this.handleDeleteClick.bind(this);
		this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
	}

	handleDeleteClick(){
		const { todo, onDeleteClick } = this.props;
		onDeleteClick(todo);
	}

	handleCheckboxClick(){
		const { todo, onCheckboxClick } = this.props;
		onCheckboxClick(todo);
	}

	render(){
		const { todo } = this.props;
		return (
			<div>
				<input checked={todo.isComplete} type="checkbox" onChange={this.handleCheckboxClick} />
				<span>{todo.label}</span>
				<span onClick={this.handleDeleteClick}> delete</span>
			</div>
		);
	}
};

Todo.propTypes = {
	todo: PropTypes.shape({
		isComplete: PropTypes.bool,
		label: PropTypes.string,
	}).isRequired,
	onCheckboxClick: PropTypes.func,
	onDeleteClick: PropTypes.func,
}

Todo.defaultProps = {
	onCheckboxClick: () => {},
	onDeleteClick: () => {},
}

export default Todo;
