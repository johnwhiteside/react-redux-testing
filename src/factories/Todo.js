import { Factory } from 'rosie';

const Todo = new Factory()
	.sequence('id')
	.sequence('label', i => `Mock Todo ${i}`)
	.attrs({ isCompleted: false });

export default Todo;
