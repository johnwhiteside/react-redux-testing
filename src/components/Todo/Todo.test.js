import React from 'react';
import Todo from './Todo';
import { shallow } from 'enzyme';

describe('Todo', () => {
	const mockTodo = {
		id: 1,
		label: 'Mock Todo',
		isComplete: false,
	}

	it('renders without error', () => {
		const props = {
			todo: mockTodo,
		}
		const wrapper = shallow(<Todo {...props} />);
		expect(wrapper.length).toEqual(1);
	});

	describe('handleDeleteClick', () => {
		it('should call onDelete', () => {
			const props = {
				todo: mockTodo,
				onDeleteClick: jest.fn(),
			}
			const wrapper = shallow(<Todo {...props} />);
			wrapper.instance().handleDeleteClick();
			expect(props.onDeleteClick).toBeCalled();
		});

		it('doesn\'t error when onDeleteClick prop is undefined', () => {
			const props = {
				todo: mockTodo,
			}
			const wrapper = shallow(<Todo {...props} />);
			expect(wrapper.instance().handleDeleteClick).not.toThrow();
		});
	});

	describe('handleCheckboxClick', () => {
		it('calls onCheckboxClick', () => {
			const props = {
				todo: mockTodo,
				onCheckboxClick: jest.fn(),
			}
			const wrapper = shallow(<Todo {...props} />);
			wrapper.instance().handleCheckboxClick();
			expect(props.onCheckboxClick).toBeCalled();
		});

		it('doesn\'t error when onCheckboxClick prop is undefined', () => {
			const props = {
				todo: mockTodo,
			}
			const wrapper = shallow(<Todo {...props} />);
			expect(wrapper.instance().handleCheckboxClick).not.toThrow();
		});
	});
});
