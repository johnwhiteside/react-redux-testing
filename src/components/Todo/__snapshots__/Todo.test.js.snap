// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo renders todo correctly 1`] = `
<div>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <span>
    Mock Todo
  </span>
  <span
    onClick={[Function]}
  >
     delete
  </span>
</div>
`;
